File: spring/src/main/java/io/github/wavesync/repository/UserResonanceNodeRepository.java
package io.github.wavesync.repository;
import io.github.wavesync.entity.UserResonanceNode;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;
import java.util.List;


@Repository
public interface UserResonanceNodeRepository extends JpaRepository<UserResonanceNode, Long> {

    @Modifying
    @Query("""
        update UserResonanceNode urn
        set urn.isDeleted = true
        where urn.userResonator.id in :ids
          and urn.isDeleted = false
    """)
    void softDeleteByUserResonatorIds(@Param("ids") List<Long> ids);

    @Query("""
        select urn
        from UserResonanceNode urn
        where urn.userResonator.id = :userResonatorId
          and urn.isDeleted = false
        order by urn.branchPosition, urn.nodePosition
    """)
    List<UserResonanceNode> findAllByUserResonatorId(Long userResonatorId);
}

File: fastapi/app/repositories/user_resonance_node_repository.py
from typing import List

from sqlalchemy import select, update
from sqlalchemy.orm import Session

from app.models.user_resonance_node import UserResonanceNode


def save_all(db: Session, nodes: List[UserResonanceNode]) -> None:
    # 커밋은 호출부 책임
    db.add_all(nodes)


def find_all_by_user_resonator_id(db: Session, user_resonator_id: int) -> List[UserResonanceNode]:
    stmt = (
        select(UserResonanceNode)
        .where(
            UserResonanceNode.user_resonator_id == user_resonator_id,
            UserResonanceNode.is_deleted.is_(False),
        )
        .order_by(UserResonanceNode.branch_position, UserResonanceNode.node_position)
    )
    return list(db.scalars(stmt))


def soft_delete_by_user_resonator_ids(db: Session, ids: List[int]) -> None:
    stmt = (
        update(UserResonanceNode)
        .where(
            UserResonanceNode.user_resonator_id.in_(ids),
            UserResonanceNode.is_deleted.is_(False),
        )
        .values(is_deleted=True)
    )
    db.execute(stmt)
